File: app/src/main/java/com/silpi/app/MainActivity.java
package com.silpi.app;

import android.os.Bundle;
import android.view.View;
import android.widget.ArrayAdapter;
import android.widget.Button;
import android.widget.CheckBox;
import android.widget.EditText;
import android.widget.Spinner;
import android.widget.TextView;
import android.widget.Toast;
import android.content.Intent;

import androidx.appcompat.app.AppCompatActivity;

public class MainActivity extends AppCompatActivity {

    Spinner spinnerCategory;
    EditText edtPeople, edtPlace, edtDistance, edtDate, edtTime, edtDesc;
    CheckBox chkGuardian;
    Button btnCreate, btnGoJoin, btnGoMap;
    TextView txtResult;

    String[] categories = {"등산", "식사", "산책", "카페", "바둑", "낚시"};

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_main);

        btnGoMap = findViewById(R.id.btnGoMap);

        btnGoMap.setOnClickListener(v -> {
            Intent intent = new Intent(MainActivity.this, MapActivity.class);
            startActivity(intent);
        });

        spinnerCategory = findViewById(R.id.spinnerCategory);
        edtPeople = findViewById(R.id.edtPeople);
        edtPlace = findViewById(R.id.edtPlace);
        edtDistance = findViewById(R.id.edtDistance);
        edtDate = findViewById(R.id.edtDate);
        edtTime = findViewById(R.id.edtTime);
        edtDesc = findViewById(R.id.edtDesc);
        chkGuardian = findViewById(R.id.chkGuardian);
        btnCreate = findViewById(R.id.btnCreate);
        txtResult = findViewById(R.id.txtResult);

        btnGoJoin = findViewById(R.id.btnGoJoin);

        btnGoJoin.setOnClickListener(v -> {
            Intent intent = new Intent(MainActivity.this, joinsilpi.class);
            startActivity(intent);
        });

        ArrayAdapter<String> adapter = new ArrayAdapter<>(
                this,
                android.R.layout.simple_spinner_item,
                categories
        );
        adapter.setDropDownViewResource(android.R.layout.simple_spinner_dropdown_item);
        spinnerCategory.setAdapter(adapter);

        btnCreate.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View view) {

                String category = spinnerCategory.getSelectedItem().toString();
                String people = edtPeople.getText().toString().trim();
                String place = edtPlace.getText().toString().trim();
                String distance = edtDistance.getText().toString().trim();
                String date = edtDate.getText().toString().trim();
                String time = edtTime.getText().toString().trim();
                String desc = edtDesc.getText().toString().trim();
                String guardian = chkGuardian.isChecked() ? "예" : "아니오";

                if (people.isEmpty() || place.isEmpty() || distance.isEmpty()
                        || date.isEmpty() || time.isEmpty()) {
                    Toast.makeText(MainActivity.this, "필수 항목을 입력하세요.", Toast.LENGTH_SHORT).show();
                    return;
                }

                // 확인창 띄우기
                new androidx.appcompat.app.AlertDialog.Builder(MainActivity.this)
                        .setTitle("모임 생성")
                        .setMessage("생성하시겠습니까?")

                        // 네 버튼
                        .setPositiveButton("네", (dialog, which) -> {

                            String result = "모임 생성 정보\n\n"
                                    + "카테고리: " + category + "\n"
                                    + "참여 인원: " + people + "명\n"
                                    + "모임 장소: " + place + "\n"
                                    + "거리: " + distance + "km\n"
                                    + "날짜: " + date + "\n"
                                    + "시간: " + time + "\n"
                                    + "보호자 동반: " + guardian + "\n"
                                    + "설명: " + desc;

                            txtResult.setText(result);

                            Toast.makeText(MainActivity.this, "모임 생성 완료", Toast.LENGTH_SHORT).show();

                            String title = category + " 모임 - " + place;

                            String newMeeting =
                                    title + "§" +
                                            category + "§" +
                                            people + "§" +
                                            place + "§" +
                                            distance + "§" +
                                            date + "§" +
                                            time + "§" +
                                            guardian + "§" +
                                            desc;

                            String oldMeetings = getSharedPreferences("meeting_data", MODE_PRIVATE)
                                    .getString("meetings", "");

                            String updatedMeetings;

                            if (oldMeetings.isEmpty()) {
                                updatedMeetings = newMeeting;
                            } else {
                                updatedMeetings = oldMeetings + "¶" + newMeeting;
                            }

                            getSharedPreferences("meeting_data", MODE_PRIVATE)
                                    .edit()
                                    .putString("meetings", updatedMeetings)
                                    .apply();
                        })

                        ///아니오 버튼
                        .setNegativeButton("아니오", null)

                        .show();
            }
        });
    }
}

File: app/src/main/java/com/silpi/app/MainActivity.kt
package com.silpi.app

import android.content.Intent
import android.os.Bundle
import android.widget.Button
import android.widget.EditText
import android.widget.Toast
import androidx.appcompat.app.AppCompatActivity
import com.google.firebase.auth.FirebaseAuth
import com.google.firebase.firestore.FirebaseFirestore

class MainActivity : AppCompatActivity() {

    private lateinit var auth: FirebaseAuth
    private lateinit var db: FirebaseFirestore

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)

        auth = FirebaseAuth.getInstance()
        db = FirebaseFirestore.getInstance()

        val currentUser = auth.currentUser
        // TODO: 이메일 인증 기능을 다시 켤 때 isEmailVerified 조건을 복구하세요.
        // if (currentUser != null && currentUser.isEmailVerified) {
        if (currentUser != null) {
            loadProfileAndMove()
            return
        }

        setContentView(R.layout.activity_main)

        val emailInput = findViewById<EditText>(R.id.emailInput)
        val passwordInput = findViewById<EditText>(R.id.passwordInput)
        val confirmButton = findViewById<Button>(R.id.confirmButton)
        val switchButton = findViewById<Button>(R.id.switchButton)
        val recoveryButton = findViewById<Button>(R.id.recoveryButton)

        confirmButton.setOnClickListener {
            val email = emailInput.text.toString().trim()
            val password = passwordInput.text.toString().trim()

            if (email.isEmpty() || password.isEmpty()) {
                Toast.makeText(this, "이메일과 비밀번호를 입력하세요", Toast.LENGTH_SHORT).show()
                return@setOnClickListener
            }

            auth.signInWithEmailAndPassword(email, password)
                    .addOnCompleteListener(this) { task ->
                        if (task.isSuccessful) {
                            val user = auth.currentUser
                            if (user == null) {
                                Toast.makeText(this, "로그인 정보를 확인할 수 없습니다.", Toast.LENGTH_SHORT).show()
                                return@addOnCompleteListener
                            }

                            user.reload().addOnCompleteListener {
                                // TODO: 이메일 인증 기능을 다시 켤 때 user.isEmailVerified 조건을 복구하세요.
                                // if (user.isEmailVerified) {
                                if (true) {
                                    Toast.makeText(this, "로그인 성공", Toast.LENGTH_SHORT).show()
                                    loadProfileAndMove()
                                } else {
                                    auth.signOut()
                                    Toast.makeText(this, "이메일 인증이 필요합니다.", Toast.LENGTH_LONG).show()
                                }
                            }
                        } else {
                            Toast.makeText(this, "이메일 또는 비밀번호 오류", Toast.LENGTH_SHORT).show()
                        }
                    }
        }

        switchButton.setOnClickListener {
            startActivity(Intent(this, SignupActivity::class.java))
        }

        recoveryButton.setOnClickListener {
            val email = emailInput.text.toString().trim()
            if (email.isEmpty()) {
                Toast.makeText(this, "이메일을 입력하세요", Toast.LENGTH_SHORT).show()
                return@setOnClickListener
            }

            auth.sendPasswordResetEmail(email)
                    .addOnCompleteListener { task ->
                        if (task.isSuccessful) {
                            Toast.makeText(this, "비밀번호 재설정 메일을 보냈습니다", Toast.LENGTH_LONG).show()
                        } else {
                            Toast.makeText(this, "메일 전송 실패: ${task.exception?.message}", Toast.LENGTH_LONG).show()
                        }
                    }
        }
    }

    private fun loadProfileAndMove() {
        val currentUser = auth.currentUser
        if (currentUser == null) {
            Toast.makeText(this, "로그인 정보를 확인할 수 없습니다.", Toast.LENGTH_SHORT).show()
            return
        }

        db.collection("users")
                .document(currentUser.uid)
                .get()
                .addOnSuccessListener { document ->
                    val user = document.toObject(User::class.java)
                    if (document.exists() && user != null) {
                        CurrentUserProvider.saveUserProfile(this, user)
                        moveToNextScreen(CurrentUserProvider.isProfileCompleted(user))
                    } else {
                        CurrentUserProvider.saveSignedInUser(
                                context = this,
                                userId = currentUser.uid,
                                userName = "",
                                email = currentUser.email.orEmpty()
                        )
                        moveToNextScreen(profileCompleted = false)
                    }
                }
                .addOnFailureListener {
                    Toast.makeText(this, "프로필 정보를 불러오지 못했습니다.", Toast.LENGTH_SHORT).show()
                    moveToProfileLoadFailed()
                }
    }

    private fun moveToProfileLoadFailed() {
        val intent = Intent(this, ProfileActivity::class.java)
                .putExtra(ProfileActivity.EXTRA_PROFILE_LOAD_FAILED, true)
        startActivity(intent)
        finish()
    }

    private fun moveToNextScreen(profileCompleted: Boolean) {
        val intent = if (profileCompleted) {
            Intent(this, HomeActivity::class.java)
        } else {
            Intent(this, ProfileActivity::class.java)
                    .putExtra(ProfileActivity.EXTRA_FIRST_SETUP, true)
        }

        startActivity(intent)
        finish()
    }
}
